kubectl delete: Add --concurrency flag

tchap opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This flag can be used to set the number of workers to use for sending delete requests to Kubernetes. By default this is set to 1 to be compatible with the current behaviour.

This change makes the deletion logic always use a queue and a worker pool, even with concurrency 1. This introduces a slight overhead, but it should not be a big deal as we spent the most time on API calls anyway.

#### Which issue(s) this PR is related to:

Fixes https://github.com/kubernetes/kubernetes/issues/125407

#### Special notes for your reviewer:

The original issue mentioned just enabling concurrency. I found it more helpful to be able to set the number of workers.

It may make sense to not use the queue and the pool on concurrency 1, but I didn't want to duplicate code.

Also it's really hard to test this in unit tests, because the fake client is not tread-safe and I didn't find a way how to hack this to make it  thread-safe. So any time you run a unit test with concurrency > 1, you get a data race error in the test. I added an e2e test, though, that doesn't really test that anything is running concurrently, but that the command works as expected and deletes the resources.

I also tested this manually.

#### Does this PR introduce a user-facing change?

```release-note
Added `--concurrency` option for `kubectl delete` that allows using multiple workers to send API requests concurrently.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

ardaguclu commented:
/ok-to-test

tchap commented:
/retest

tchap commented:
/retest

lmktfy commented:
Changelog suggestion
```diff
-Added kubectl delete --concurrency flag for using multiple workers for sending API requests concurrently
+Added `--concurrency` option for `kubectl delete` that allows using multiple workers to send API requests concurrently.
```

tchap committed:
kubectl delete: Add --concurrency flag

This flag can be used to set the number of workers to use for sending
delete requests to Kubernetes. By default this is set to 1 to be
compatible with the current behaviour.

This change makes the deletion logic always use a queue and a worker
pool, even with concurrency 1. This introduces a slight overhead, but it
should not be a big deal as we spent the most time on API calls anyway.
